### DISCUSSION TITLE
Firebase middleware for authentication

### DISCUSSION THREAD

**Discussion by vimtor** [vercel/next.js#33586] [Help] [answered]
Hi there!

I planned on using Firebase + Next.js Middlewares to check for user authentication.
But when I try to do something like this:

```ts
import type { NextRequest } from "next/server";
import { NextResponse } from "next/server";
import { auth } from "../../lib/firebase/admin";

export async function middleware(req: NextRequest) {
  if (req.nextUrl.href.startsWith("/api/login")) {
    const response = NextResponse.next();

    const cookie = await auth.createSessionCookie(req.headers.get("token"), {
      expiresIn: 9999999,
    });

    response.cookie("token", cookie, {
      httpOnly: true,
      sameSite: "strict",
      secure: process.env.NODE_ENV === "production",
    });
  }

  const user = await auth.verifySessionCookie(req.cookies.token);
  return NextResponse.next();
}
```

I get the following error:

<img width="1028" alt="CleanShot 2022-01-23 at 22 26 20@2x" src="https://user-images.githubusercontent.com/36263538/150698601-c30d5e3e-a8d0-4cf4-90e6-897377b0fe74.png">

I've tried to install `http2` but other old packages (`dns`, `net`, `tls`) keep popping up with the same error.
Is the Firebase admin SDK incompatible with Next.js middleware?

**Comment by wwwdepot** [comment, marked as answer]
Yep, middleware doesn't support Node.js API. See https://nextjs.org/docs/api-reference/edge-runtime

**Reply by vukadinFE** [reply]
But why?

**Reply by vimtor** [reply]
This is quite sad because it prevent many libraries to be used as middlewares. In fact the middleware example repo does not have many popular tools. I would be really cool to know why this decision was made and if this is planned to be supported.

**Comment by awinogrodzki** [comment]
Hey, I've recently released a library that aims to solve the problem: https://github.com/ensite-in/next-firebase-auth-edge

It allows to create and verify tokens inside Next.js middleware and Next.js 13 server components. Built entirely upon Web Crypto API

Follow the link for some code examples.

**Reply by namnn9x** [reply]
how can i verifySessionCookie when using your library

**Reply by awinogrodzki** [reply]
You can use [authentication middleware](https://github.com/awinogrodzki/next-firebase-auth-edge#authentication-middleware).

It automatically sets and validates cookies.

**Comment by AmitayCohen1** [comment]
As @awinogrodzki noted, the "next-firebase-auth-edge" library addresses the Edge runtime <> Firebase issue. If you're interested, I wrote a guide on authenticating a NextJS app with Firebase while preserving SSR features:
https://gitgit.substack.com/p/nextjs-authentication-with-firebase
